### PR TITLE
Pool HitObjects, explosions and judgements

### LINKED ISSUE #25: Minibosses do not show judgements

### LINKED ISSUE #59: Pool hit explosions
We should aim for zero allocations.

### PR DESCRIPTION
**Pull request by LumpBloom7**
* [x] Depends on #131 
* [x] Extra adjustments to make full use of #131 

Resolves #59
Fixes #25 

I've made pooling work for each HitObject and nested HitObjects, along with explosions, judgements, and the health text. I tried to not touch anything irrelevant or do anything that I would consider out of scope for this PR.

## Encountered problems: 

~~Some objects and properties aren't lane agnostic, while most objects are fine in this regard, I find the anchor adjustments to be a bit ugly. `MinionPiece` also changes sprites based on the lane the DHO is in, and [I've made it so that it updates its animation set when the DHO has a new applied HitObject](https://github.com/LumpBloom7/rush/blob/4a3484334abd1418d2eec9ab77c0ba3ef35ca99c/osu.Game.Rulesets.Rush/Objects/Drawables/Pieces/MinionPiece.cs#L39). These issues, as mentioned earlier, could be completely alleviated with nested playfields.~~ With the use of nested playfields, we now preassign a lane to the newly created DHO, so that the piece will load in the proper textures. This DHO will always be used in the same lane, so we don't need to reapply the visuals on every use now.

Next problem I couldn't really get my head around. The `MinionPiece` occasionally appears invisible when proxying is used within `DrawableMinion`, and very inconsistent at that. The `TextureAnimation` drawable is being masked for some reason, and that is worth investigating imo. I've chosen to comment out the proxying code, since that keeps everything functional for now. The whole proxying code can also be removed imo seeing as the HitObjectContainer is reverse-ordered, and that the proxy layer is right on top of HOC, rendering it a redundant procedure.

~~I'm also not happy with how I went about with dealing with the non-DHO poolables, mainly the various explosions. I wanted to centralize the initialization of each to the `Apply()` methods, but some had a dependance of a parent's DrawSize. Some of the transforms are also not lane agnostic, meaning some more transforms are added to ensure the proper state before the real-deal. (otherwise a judgement text may animate in a different lane)~~ Most of these are now lane agnostic, the lane is responsible for their positioning now.

### CODE REVIEW

**Comment by frenzibyte** [comment]
I've done a thorough review and have pushed preliminary changes for this, looks pretty good. @swoolcock requesting your review and merge.

**Review by frenzibyte — APPROVED** [review]

File: osu.Game.Rulesets.Rush/UI/RushPlayfield.cs [hunk 1]
```diff
@@ -161,29 +170,45 @@ public RushPlayfield()
...
-                else if (laned.Lane == LanedHitLane.Ground)
-                    groundLane.Add(hitObject);
+        public override void Add(DrawableHitObject hitObject)
+        {
+            if (!(hitObject is IDrawableLanedHit laned))
                 return;
-            }
 
-            base.Add(hitObject);
+            if (laned.Lane == LanedHitLane.Air)
```
Comment (swoolcock) [inline osu.Game.Rulesets.Rush/UI/RushPlayfield.cs:208]: Use playfieldForLane?

**Review by swoolcock — CHANGES_REQUESTED** [review]
Just one small thing.

**Comment by swoolcock** [comment]
LGTM

**Pull request merged**